### PR TITLE
KAA-1231 [Java] Client authentication key pair generation runtime paramether

### PR DESCRIPTION
**Pull request by maxml**
http://jira.kaaproject.org/browse/KAA-1231

### CODE REVIEW

**Comment by kaa-leeroy** [comment]
Can one of the admins verify this patch?

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/Kaa.java [hunk 1]
```diff
@@ -16,21 +16,20 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+
 /**
  * Creates new Kaa client based on {@link KaaClientPlatformContext platform
  * context} and optional {@link KaaClientStateListener state listener}.
- *
```
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/Kaa.java:31]: Missing description of <b>isUserKeyStrategy</b> parameter, and why it's named <b>isUserKeyStrategy</b> ?

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java [hunk 2]
```diff
@@ -376,14 +391,17 @@ private KeyPair getOrInitKeyPair() {
...
+            throw new KaaRuntimeException("Key pair is not found and your key strategy is default");
         }
         return keyPair;
     }
 
     @Override
     public EndpointKeyHash getEndpointKeyHash() {
         if (keyHash == null) {
-            EndpointObjectHash publicKeyHash = EndpointObjectHash.fromSHA1(getOrInitKeyPair().getPublic().getEncoded());
+            EndpointObjectHash publicKeyHash = EndpointObjectHash.fromSHA1(getOrInitKeyPair(false).getPublic().getEncoded());
```
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java]: Why <b>false</b> value is used here instead of <b>isUserKeyStrategy</b>

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientState.java [hunk 3]
```diff
@@ -31,6 +32,11 @@
     boolean isRegistered();
     void setRegistered(boolean registered);
 
+    /**
+     * User can choose to create key pair or not. In default case there is no key generation
+     *
```
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientState.java]: Description of key generation strategies should be for <b>isUserKeyStrategy</b> parameter. The methods below don't have any arguments and just return keys. Description specifying what kind of key is returned should be added individually to each method.

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java (hunk 2 above)
Comment (maxml) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java]: because of not create key pair, but use already created

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/Kaa.java [hunk 4]
```diff
@@ -43,14 +42,28 @@ public static KaaClient newClient(KaaClientPlatformContext context) throws KaaRu
...
     public static KaaClient newClient(KaaClientPlatformContext context, KaaClientStateListener listener) throws KaaRuntimeException {
+        return newClient(context, listener, false);
+    }
+
+    /**
+     * Creating new Kaa client
+     *
+     * @param context           - Kaa platform specific context for client init
+     * @param listener          - notifier for changing client state
+     * @param isUserKeyStrategy - user strategy about creating public/private keys. In default case use previously created
```
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/Kaa.java]: The name <b>isUserKeyStrategy</b> is confusing. I think it should be named <b>autogenerateKeys</b>. Please add to the description meaning of its values. 'false' - use existing pre-generated public/private key pair (default behaviour). 'true' - auto-generate new public/private key pair if absent.

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java (hunk 2 above)
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java]: I insist on using <b>autogenerateKeys</b> variable here since we can't guarantee that _getOrIinitKeyPair_ won't be invoked here first time.

File: pom.xml [hunk 5]
```diff
@@ -30,7 +30,9 @@ Copyright 2014-2016 CyberVision, Inc.
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
         <main.dir>${basedir}</main.dir>
-        <akka.version>2.4.1</akka.version>
+        <avro-ui.version>0.3.0-SNAPSHOT</avro-ui.version>
```
Comment (ikulikov) [inline pom.xml]: What is the reason of changing avro-ui version here?

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/bootstrap/BootstrapManager.java [hunk 6]
```diff
@@ -16,26 +16,26 @@
 
 package org.kaaproject.kaa.client.bootstrap;
 
-import java.util.List;
-
```
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/bootstrap/BootstrapManager.java:20]: There are a lot of files with changed formatting which are unrelated to this task. Please revert formatting for such files in order to have changes only related to this task.

Comment (maxml) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/bootstrap/BootstrapManager.java:20]: I don't know, how I can avoid this. I think formatting help to read code faster and maintain t faster. I use only IDE formatting. I understand, that hard to read, but don't know how I can help in this. Don't want to point to smth, but maybe we can create some formatting rules, one configuration for all? if formatting in this project have some special requirements, we can format all projects to one style. It will be very helpful, as I can see this...

Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/bootstrap/BootstrapManager.java:20]: I'm not against code style. I mean that code formatting is out of scope these changes. Code style can applied in scope of another specific ticket.

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/Kaa.java [hunk 7]
```diff
@@ -43,14 +42,30 @@ public static KaaClient newClient(KaaClientPlatformContext context) throws KaaRu
...
     public static KaaClient newClient(KaaClientPlatformContext context, KaaClientStateListener listener) throws KaaRuntimeException {
+        return newClient(context, listener, false);
+    }
+
+    /**
+     * Creating new Kaa client
+     *
+     * @param context             - Kaa platform specific context for client init
+     * @param listener            - notifier for changing client state
+     * @param isAutogeneratedKeys - check if creating public/private key pair.
```
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/Kaa.java]: isAutogeneratedKeys - whether to generate public/private key pair automatically.

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java (hunk 2 above)
Comment (ikulikov) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/KaaClientPropertiesState.java]: My recommendation to use  <b>isAutogeneratedKeys</b> variable is still ignored.
